### ISSUE TITLE
DNS Proxy is not able to resolve the address of a ServiceEntry using workloadSelector

### ISSUE THREAD

**Issue by ferrarimarco** [istio/istio#33191]
**Bug description**

TLDR: DNS Proxy is not able to resolve the address of a ServiceEntry, if I use `workloadSelector`. The DNS proxy resolves the address if I manually specify the endpoints.

I deployed a workload on a VM (the bookinfo example app), added the VM to the mesh, and applied a `WorkloadEntry` and the corresponding `ServiceEntry` to the namespace. According to [DNS Proxy](https://istio.io/latest/docs/ops/configuration/traffic-management/dns-proxy/), this should work:

> While Kubernetes provides DNS resolution for Kubernetes Services out of the box, any custom ServiceEntrys will not be recognized. With this feature, ServiceEntry addresses can be resolved without requiring custom configuration of a DNS server. For Kubernetes Services, the DNS response will be the same, but with reduced load on kube-dns and increased performance.

To investigate the issue, I deployed an instance of the `sleep` sample application, and followed the steps in [Debugging Envoy and Istiod](https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/). I see the endpoint marked as `HEALTHY`, so I assume that the corresponding service is available in the service registry.

I don't see any errors in istiod logs.

```
$ istioctl proxy-status
NAME                                                    CDS        LDS        EDS        RDS          ISTIOD                      VERSION
istio-eastwestgateway-654c6c97f6-pf265.istio-system     SYNCED     SYNCED     SYNCED     NOT SENT     istiod-56874696b5-njjkw     1.10.0
istio-egressgateway-585f7668fc-68q7s.istio-system       SYNCED     SYNCED     SYNCED     NOT SENT     istiod-56874696b5-njjkw     1.10.0
istio-ingressgateway-8657768d87-dp4lt.istio-system      SYNCED     SYNCED     SYNCED     SYNCED       istiod-56874696b5-njjkw     1.10.0
legacy-vm.default                                       SYNCED     SYNCED     SYNCED     SYNCED       istiod-56874696b5-njjkw     1.10.0
sleep-5bc78c5489-drszd.default                          SYNCED     SYNCED     SYNCED     SYNCED       istiod-56874696b5-njjkw     1.10.0

$ istioctl proxy-config endpoint sleep-5bc78c5489-drszd.default --cluster "outbound|9080||productpage.default.svc.cluster.local"
ENDPOINT               STATUS      OUTLIER CHECK     CLUSTER
34.68.112.221:9080     HEALTHY     OK                outbound|9080||productpage.default.svc.cluster.local
```

Here are the descriptors I'm using:

```yaml
---
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
spec:
  components:
    ingressGateways:
      - name: istio-eastwestgateway
        label:
          istio: eastwestgateway
          app: istio-eastwestgateway
        enabled: true
        k8s:
          env:
            # sni-dnat adds the clusters required for AUTO_PASSTHROUGH mode
            - name: ISTIO_META_ROUTER_MODE
              value: "sni-dnat"
          service:
            ports:
              - name: status-port
                port: 15021
                targetPort: 15021
              - name: tls
                port: 15443
                targetPort: 15443
              - name: tls-istiod
                port: 15012
                targetPort: 15012
              - name: tls-webhook
                port: 15017
                targetPort: 15017
  meshConfig:
    defaultConfig:
      proxyMetadata:
        # Enable basic DNS proxying
        ISTIO_META_DNS_CAPTURE: "true"
        # Enable automatic address allocation
        ISTIO_META_DNS_AUTO_ALLOCATE: "true"
  values:
    global:
      meshID: mesh1
      multiCluster:
        clusterName: "Kubernetes"
      network: ""
...

---
apiVersion: networking.istio.io/v1alpha3
kind: WorkloadEntry
metadata:
  name: productpage
spec:
  address: [IP ADDRESS OF THE VM]
  labels:
    app: productpage
    instance-id: source-environment-vm
  serviceAccount: bookinfo-gce
...
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: productpage
spec:
  hosts:
    - productpage.default.svc.cluster.local
  location: MESH_INTERNAL
  ports:
    - number: 9080
      name: http
      protocol: HTTP
  resolution: STATIC
  workloadSelector:
    labels:
      app: productpage
...
---
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: dns-proxy-test
spec:
  addresses:
    - 1.2.3.4
  hosts:
    - dns.proxy.test
  ports:
    - name: http
      number: 80
      protocol: HTTP
...
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: productpage
spec:
  host: productpage.default.svc.cluster.local
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
...
```

Output from the VM:

```
$ curl -v productpage.default.svc.cluster.local
* Rebuilt URL to: productpage.default.svc.cluster.local/
* Could not resolve host: productpage.default.svc.cluster.local
* Closing connection 0
curl: (6) Could not resolve host: productpage.default.svc.cluster.local

curl -v dns.proxy.test
* Rebuilt URL to: dns.proxy.test/
*   Trying 1.2.3.4...
```

Output from the sleep Pod:

```
$ curl -v productpage.default.svc.cluster.local
* Could not resolve host: productpage.default.svc.cluster.local
* Closing connection 0
curl: (6) Could not resolve host: productpage.default.svc.cluster.local

$ curl -v dns.proxy.test
*   Trying 1.2.3.4:80...
* Connected to dns.proxy.test (1.2.3.4) port 80 (#0)
> GET / HTTP/1.1
> Host: dns.proxy.test
> User-Agent: curl/7.77.0-DEV
> Accept: */*
>
```

[x] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

Both the Pod and the VM should be able to resolve the IP address of the ServiceEntry via DNS proxy.

**Steps to reproduce the bug**

1. Enable DNS proxy.
2. Apply the above descriptors.
3. Try resolving `productpage.default.svc.cluster.local` and `dns.proxy.test`

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

```
$ istioctl version --remote
client version: 1.10.0
control plane version: 1.10.0
data plane version: 1.10.0 (5 proxies)

$ kubectl version --short
Client Version: v1.21.1
Server Version: v1.20.6-gke.1000

```

**How was Istio installed?**

Via the Istio operator

**Environment where the bug was observed (cloud vendor, OS, etc)**

Google Kubernetes Engine

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by howardjohn** [comment]
To clarify, are you  expecting `productpage.default.svc.cluster.local` to resolve to `[IP ADDRESS OF THE VM]`? Or to some random IP due to `ISTIO_META_DNS_AUTO_ALLOCATE=true`?

**Comment by ferrarimarco** [comment]
To some random IP, actually.

I completely re-initialized the cluster, installed Istio, and now it works from both the Pod and the VM.

```
curl -v productpage.default.svc.cluster.local
* Rebuilt URL to: productpage.default.svc.cluster.local/
*   Trying 240.240.0.8...
```

For some reason, I can't reproduce it anymore.

Closing for now. Will re-open if I encounter this again.

Thanks for your support mate!
